File: quicksort/c++/quicksort.cpp
/*
    quicksort implementation
    1. partition function
    2. quicksort function

    Time complexity, Worst-case ==>> O(n^2)
                Average-Expected ==>> O(n*logn)
    Space complexity, Big O ==>> O(1), in-place algorithm
*/ 

#include <bits/stdc++.h>

using namespace std;

int partition(int A[], int p, int r){
    int pivot = A[r];
    int i = p-1;
    for (int j = p; j < r; j++){
        if(A[j]<pivot){
            i++;
            swap(A[j], A[i]);
        }
    }
    swap(A[i+1], A[r]);
    return i+1;
}

void quicksort(int A[], int p, int r){
    if(p<r){
        int q = partition(A, p, r);
        quicksort(A, p, q-1);
        quicksort(A, q+1, r);
    }
}

int main(){
    // example of usage
    // let's take the following an array
    int A[] = {6, 4, 1, 23, 46, -3, 12};
    // calculate its length
    int array_size = sizeof(A)/sizeof(A[0]);
    // apply quicksort
    quicksort(A, 0, array_size - 1);
    // check the result
    for (size_t i = 0; i < array_size; i++){
        cout << A[i] << " ";
    }
    cout << endl;
}

File: quicksort/python/quicksort.py
# quicksort implementation
# 1. partition function
# 2. quicksort function

# Time complexity, Worst-case ==>> O(n*logn)
#            Average-Expected ==>> O(n*logn)
# Space complexity, Big O ==>> O(1), in-place algorithm

def partition(A, p, r):
    """
    implementation of partition function used for quicksort
    """
    pivot = A[r]    # select pivot element
    i = p - 1
    for j in range(p, r):
        if A[j]<pivot:      # if an element is less than a pivot, exchange i and jth elements
            i+=1
            A[j], A[i] = A[i], A[j]
    A[r], A[i+1] = A[i+1], pivot        # at the end exchange pivot with i+1 th element of the list
    return i+1      # return the index where partition should be performed

def quicksort(A, p, r):
    """
    implementation of quicksort
    """
    if p<r:
        q = partition(A, p, r)      # select an index to apply partition
        quicksort(A, p, q-1)        # apply quicksort for left side
        quicksort(A, q+1, r)        # apply quicksort for right side

# Example
A = [9,7, 4, 6, 2, 5]
quicksort(A, 0, len(A)-1)
print(A)
# which prints: [2, 4, 5, 6, 7, 9]
